device_doctor/lib/src/adb.dart

```diff
@@ -0,0 +1,397 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert' show LineSplitter;
+import 'dart:io';
+
+import 'package:meta/meta.dart';
+
+import 'utils.dart';
+
+/// The root of the API for controlling devices.
+DeviceDiscovery get devices => DeviceDiscovery();
+
+/// Type of host that is configured to test.
+enum HostType { vm, physical }
+
+/// Operating system on the devices that this host is configured to test.
+enum DeviceOperatingSystem { android, ios, none }
+
+/// Discovers available devices and chooses one to work with.
+abstract class DeviceDiscovery {
+  factory DeviceDiscovery() {
+    switch (config.deviceOperatingSystem) {
+      case DeviceOperatingSystem.android:
+        return AndroidDeviceDiscovery();
+      case DeviceOperatingSystem.ios:
+        return IosDeviceDiscovery();
+      case DeviceOperatingSystem.none:
+        return NoOpDeviceDiscovery();
+      default:
+        throw StateError('Unsupported device operating system: {config.deviceOperatingSystem}');
+    }
+  }
+
+  /// Lists all available devices' IDs.
+  Future<List<Device>> discoverDevices({int retriesDelayMs = 10000});
+
+  /// Checks the health of the available devices.
+  Future<Map<String, HealthCheckResult>> checkDevices();
+
+  /// Restarts the device.
+  Future<void> restartDevice();
+}
+
+/// A proxy for one specific device.
+abstract class Device {
+  /// A unique device identifier.
+  String get deviceId;
+
+  /// Whether the device is awake.
+  Future<bool> isAwake();
+
+  /// Whether the device is asleep.
+  Future<bool> isAsleep();
+
+  /// Wake up the device if it is not awake.
+  Future<void> wakeUp();
+
+  /// Send the device to sleep mode.
+  Future<void> sendToSleep();
+
+  /// Emulates pressing the power button, toggling the device's on/off state.
+  Future<void> togglePower();
+
+  /// Turns off TalkBack on Android devices, does nothing on iOS devices.
+  Future<void> disableAccessibility();
+
+  /// Unlocks the device.
+  ///
+  /// Assumes the device doesn't have a secure unlock pattern.
+  Future<void> unlock();
+
+  /// Restarts the device.
+  Future<void> restart();
+}
+
+/// Constant battery health values returned from Android Battery Manager.
+///
+/// https://developer.android.com/reference/android/os/BatteryManager.html
+class AndroidBatteryHealth {
+  // Match SCREAMING_CAPS to Android constants.
+  static const BATTERY_HEALTH_UNKNOWN = 1;
+  static const BATTERY_HEALTH_GOOD = 2;
+  static const BATTERY_HEALTH_OVERHEAT = 3;
+  static const BATTERY_HEALTH_DEAD = 4;
+  static const BATTERY_HEALTH_OVER_VOLTAGE = 5;
+  static const BATTERY_HEALTH_UNSPECIFIED_FAILURE = 6;
+  static const BATTERY_HEALTH_COLD = 7;
+}
+
+class AndroidDeviceDiscovery implements DeviceDiscovery {
```

Comment:
The flutter command line tool has device discovery already built in. Is it possible for you to re-use it?

Reply:
Unfortunately the device management tool will be called before task (recipes), where the flutter tool is not ready yet.